### PR TITLE
Writing more specific specs and documentation

### PR DESCRIPTION
**Pull request by icidasset**
Thoughts?

### CODE REVIEW

File: lib/exsom.ex [hunk 1]
```diff
@@ -12,8 +12,23 @@ defmodule Exsom do
...
+  @type binary_xsd    :: binary       # XSD schema in binary form
+  @type charlist_xsd  :: charlist     # XSD schema in charlist form
+  @type decoded_xsd   :: [integer]    # XSD schema represented by unicode code points
+
+  @type xsd           :: decoded_xsd | charlist_xsd | binary_xsd
+  @type raw_xsd       :: decoded_xsd | binary_xsd
+
+  @typedoc "Internal representation of a XSD document"
+  @type model         :: any
+
```
Comment (expede) [inline lib/exsom.ex]: 👍 🎉

Comment (expede) [inline lib/exsom.ex]: If `decoded_xml` can be a raw string, how is it decoded? I think that I'm confused about something

File: lib/exsom/xsd.ex [hunk 2]
```diff
@@ -1,11 +1,12 @@
 defmodule Exsom.XSD do
   @moduledoc ~S"""
-  Functions for reading and writing XSD (schema) files
+  Functions for handling in-memory XSD documents.
```
Comment (expede) [inline lib/exsom/xsd.ex]: Much clearer; thanks 😄

File: lib/exsom.ex (hunk 1 above)
Comment (icidasset) [inline lib/exsom.ex]: Yeah, good point. I guess that is still confusing, I took that from erlsom docs:

> XSD can be an encoded binary (see section on character encoding) or a decoded list of Unicode code points.

Comment (icidasset) [inline lib/exsom.ex]: Naming things 😖 😂

**Comment by expede** [comment]
Looking good so far! Thanks for updating the typespecs 😅

File: lib/exsom.ex (hunk 1 above)
Comment (expede) [inline lib/exsom.ex]: I guess it's for times when people want to work with the integer representation? Different strokes for different folks, I guess 😛

🚢

Comment (icidasset) [inline lib/exsom.ex]: Hehe yeah. I tried to think of a word that represents it better (ie. a list of integers), but I can't think of anything decent...

Comment (expede) [inline lib/exsom.ex]: Maybe just a note about the use case would clarify for readers of the docs down the road (ie: us in 3 weeks 😉). Even just

```elixir
@type decoded_xml   :: [integer]    # XML document represented by Unicode code points (ie: a list of integers)
```

**Comment by icidasset** [comment]
@expede I started writing some tests and turns out that `:erlsom.write(struct, model, options)` (ie. `write/3`) doesn't work at all, but `write/2` does (without the options). I don't get it, the erlsom code clearly exports both methods 😕 I'll look into this more tomorrow, but if you have any ideas, please send them my way 😄

**Comment by expede** [comment]
@icidasset what the heck? I'll take a look in a bit, but that's hella weird. To narrow the issue, does `:earldom.write/3` work as expected?

**Comment by icidasset** [comment]
Uhm, not sure what you mean? `:erlsom.write/3` doesn't work, `:erlsom.write/2` does.

**Comment by expede** [comment]
Ah, sorry. Looking at the Erlsom code, it seems that `:erlsom.write(struct, model) == :erlsom.write(struct, model, [])`. Does `:erlsom.write(struct, model, [])` work?

**Comment by icidasset** [comment]
Nope 😢 

```
1) test integration (ExsomTest)
     test/exsom_test.exs:5
     ** (UndefinedFunctionError) function :erlsom.write/3 is undefined or private. Did you mean one of:
     
           * write/2
     
     stacktrace:
       :erlsom.write({:foo_type, [], 'yo', ['1']}, {:model, [{:type, :_document, :sequence, [{:el, [{:alt, :foo, :foo_type, [], 1, 1, true, :undefined}], 1, 1, :undefined, 3}], [], :undefined, :undefined, 1, 1, 1, false, :undefined}, {:type, :foo_type, :sequence, [{:el, [{:alt, :bar, {:"#PCDATA", :char}, [], 1, 1, true, :undefined}], 1, :unbound, :undefined, 4}], [{:att, :attr, 3, true, :char}], :undefined, :undefined, 3, 1, 1, :undefined, :undefined}], [{:ns, 'http://www.w3.org/2001/XMLSchema', 'xsd', :qualified}], :undefined, [], true, :skip}, [])
       test/exsom_test.exs:9: (test)
```

**Comment by expede** [comment]
Okay, something is seriously effed then. Looking into it

**Comment by expede** [comment]
Found it!

**Comment by expede** [comment]
We're using Erlsom 1.4.1 (the version from Hex), that [does not expose `write/3`](https://github.com/willemdj/erlsom/blob/v1.4.1/src/erlsom.erl#L37). If we need `write/3`, we need to use `master`.

**Comment by icidasset** [comment]
Omg, really, I checked the version from master and it was also `1.4.1`, so I thought it was the same. I should have known better 👎  Thanks!! 🎉 I'll use the `master` branch then.

**Comment by expede** [comment]
I should probably be more specific, `master@HEAD` is `9b3f9106a9b1376ab7a4dbdd11cae7bb9835ef0a`, whereas `.1.4.1` is `7640eb0915e71013e70c3f7bac1291262a520d96`

**Comment by icidasset** [comment]
I changed it up a bit more and wrote some more tests (not much though). Probably going to leave it at this for the moment, I have to learn about the others parts in order to write documentation for it (ie. SAX, utf8, utf16, etc). Thoughts?

File: mix.exs [hunk 3]
```diff
@@ -28,7 +28,7 @@ defmodule Exsom.Mixfile do
 
   defp deps do
     [{:earmark, "~> 0.1", only: :dev},
-     {:erlsom, "~> 1.4"},
+     {:erlsom, git: "https://github.com/willemdj/erlsom.git"},
```
Comment (expede) [inline mix.exs]: We may want to set this to the current commit, so that it doesn't break if they make a major change on `master`

**Comment by expede** [comment]
Just the one comment, but otherwise: ship it!

**Comment by icidasset** [comment]
🎉 Thanks! I will add the commit hash to the depedency.

**Pull request merged**